### PR TITLE
[rpc] Allow RPC to fetch all addrman records and add records to addrman

### PR DESCRIPTION
**Pull request by jnewbery**
Currently addrman only allows a maximum of 1000 records or 23% of all records to be returned in a call to `GetAddr()`. Relax this limit and have the client specify the max records they want. For p2p, behaviour is unchanged (but the rate limiting is set inside net_processing, where it belongs). For RPC, `getnodeaddresses` can now return the complete addrman, which is helpful for testing and monitoring.

Also add a test-only RPC `addpeeraddress`, which adds an IP address:port to addrman. This is helpful for testing (eg #18991).

### CODE REVIEW

**Comment by jnewbery** [comment]
cc @naumenkogs

**Comment by naumenkogs** [comment]
I think this is a move in the right direction: GetAddr() should be more dynamic.
It also seems to be useful for testing.

Concept ACK.

**Comment by laanwj** [comment]
Concept ACK, I think more control over addrman via RPC is a good thing.

**Review by mzumsande — COMMENTED** [review]
Concept ACK, the getaddr limits only make sense for P2P, not for RPC.

File: src/rpc/net.cpp [hunk 1]
```diff
@@ -773,6 +773,54 @@ static UniValue getnodeaddresses(const JSONRPCRequest& request)
...
+            },
+        },
+        RPCExamples{
+            HelpExampleCli("addpeeraddress", "\"1.2.3.4\" 8333")
+    + HelpExampleRpc("getnodeaddresses", "\"1.2.3.4\", 8333")
+        },
+    }.Check(request);
+
+    NodeContext& node = EnsureNodeContext(request.context);
+    if (!node.addrman) {
```
[the remark this answers is not in the document]
Comment (jnewbery) [reply src/rpc/net.cpp]: oops. This was from a previous version of the branch. Should be fixed now.

File: src/rpc/net.cpp [hunk 2]
```diff
@@ -775,6 +773,54 @@ static UniValue getnodeaddresses(const JSONRPCRequest& request)
...
+
+    NodeContext& node = EnsureNodeContext(request.context);
+    if (!node.connman) {
+        throw JSONRPCError(RPC_CLIENT_P2P_DISABLED, "Error: Peer-to-peer functionality missing or disabled");
+    }
+
+    UniValue obj(UniValue::VOBJ);
+
+    std::string addr_string = request.params[0].get_str();
+    uint16_t port = request.params[1].get_int();
```
Comment (mzumsande) [inline src/rpc/net.cpp:804]: Should add `addpeeraddress` to `vRPCConvertParams` (`rpc/client.cpp`).

File: src/rpc/net.cpp [hunk 3]
```diff
@@ -775,6 +773,54 @@ static UniValue getnodeaddresses(const JSONRPCRequest& request)
...
+        },
+        RPCResult{
+            RPCResult::Type::OBJ, "", "",
+            {
+                {RPCResult::Type::BOOL, "success", "whether the peer address waas successfully added to the address manager"},
+            },
+        },
+        RPCExamples{
+            HelpExampleCli("addpeeraddress", "\"1.2.3.4\" 8333")
+    + HelpExampleRpc("getnodeaddresses", "\"1.2.3.4\", 8333")
```
Comment (mzumsande) [inline src/rpc/net.cpp:792]: typo, getnodeaddresses->addpeeraddress

File: src/rpc/net.cpp [hunk 4]
```diff
@@ -775,6 +773,54 @@ static UniValue getnodeaddresses(const JSONRPCRequest& request)
...
+    RPCHelpMan{"addpeeraddress",
+        "\nAdd the address of a potential peer to the address manager. This RPC is for testing only.\n",
+        {
+            {"address", RPCArg::Type::STR, RPCArg::Optional::NO, "The IP address of the peer"},
+            {"port", RPCArg::Type::NUM, RPCArg::Optional::NO, "The port of the peer"},
+        },
+        RPCResult{
+            RPCResult::Type::OBJ, "", "",
+            {
+                {RPCResult::Type::BOOL, "success", "whether the peer address waas successfully added to the address manager"},
```
Comment (mzumsande) [inline src/rpc/net.cpp:787]: typo: waas -> was

File: src/net.h [hunk 5]
```diff
@@ -51,11 +51,8 @@ static const bool DEFAULT_WHITELISTFORCERELAY = false;
 static const int TIMEOUT_INTERVAL = 20 * 60;
 /** Run the feeler connection loop once every 2 minutes or 120 seconds. **/
 static const int FEELER_INTERVAL = 120;
-/** The maximum number of new addresses to accumulate before announcing. */
-static const unsigned int MAX_ADDR_TO_SEND = 1000;
-// TODO: remove ADDRMAN_GETADDR_MAX and let the caller specify this limit with MAX_ADDR_TO_SEND.
-static_assert(MAX_ADDR_TO_SEND == ADDRMAN_GETADDR_MAX,
-    "Max allowed ADDR message size should be equal to the max number of records returned from AddrMan.");
+/** The maximum number of addresses from our addrman to return in a getaddr call. */
```
Comment (mzumsande) [inline src/net.h:54]: "call" still sounds like rpc, maybe "return in response to a getaddr message".

File: src/net_processing.cpp [hunk 6]
```diff
@@ -143,6 +143,8 @@ static constexpr unsigned int MAX_FEEFILTER_CHANGE_DELAY = 5 * 60;
 static constexpr uint32_t MAX_GETCFILTERS_SIZE = 1000;
 /** Maximum number of cf hashes that may be requested with one getcfheaders. See BIP 157. */
 static constexpr uint32_t MAX_GETCFHEADERS_SIZE = 2000;
+/** the maximum percentage of addresses from our addrman to return in a getaddr call. */
```
Comment (mzumsande) [inline src/net_processing.cpp:146]: as for "call", see above

File: test/functional/rpc_net.py [hunk 7]
```diff
@@ -154,30 +152,35 @@ def test_service_flags(self):
...
-        node_addresses = self.nodes[0].getnodeaddresses(REQUEST_COUNT)
-        assert_equal(len(node_addresses), REQUEST_COUNT)
+            self.nodes[0].addpeeraddress(a, 8333)
+
+        # Obtain addresses via rpc call and check they were ones sent in before.
+        #
+        # All addresses added above are in the same netgroup and so are assigned
+        # to the same bucket. Maximum possible addresses in addrman is therefore
+        # 64, although actual number will usually be slightly less due to
+        # BucketPosition collisions.
```
Comment (mzumsande) [inline test/functional/rpc_net.py:171]: This comment became outdated.

**Review by mzumsande — COMMENTED** [review]
I did a bit of testing via console, and everything worked as expected (after adding `addpeeraddress` to `vRPCConvertParams`).
Added some minor issues/nits.

File: src/rpc/net.cpp (hunk 2 above)
Comment (jnewbery) [reply src/rpc/net.cpp:804]: done

File: src/rpc/net.cpp (hunk 3 above)
Comment (jnewbery) [reply src/rpc/net.cpp:792]: fixed

File: src/rpc/net.cpp (hunk 4 above)
Comment (jnewbery) [reply src/rpc/net.cpp:787]: fixed

File: src/net.h (hunk 5 above)
Comment (jnewbery) [reply src/net.h:54]: fixed

File: src/net_processing.cpp (hunk 6 above)
Comment (jnewbery) [reply src/net_processing.cpp:146]: fixed

File: test/functional/rpc_net.py (hunk 7 above)
Comment (jnewbery) [reply test/functional/rpc_net.py:171]: fixed

**Comment by jnewbery** [comment]
Thanks for the review @mzumsande . All comments have been addressed.

File: src/addrman.cpp [hunk 8]
```diff
@@ -479,11 +479,15 @@ int CAddrMan::Check_()
...
 #endif
 
-void CAddrMan::GetAddr_(std::vector<CAddress>& vAddr)
+void CAddrMan::GetAddr_(std::vector<CAddress>& vAddr, unsigned int max_addresses, unsigned int max_pct)
 {
-    unsigned int nNodes = ADDRMAN_GETADDR_MAX_PCT * vRandom.size() / 100;
-    if (nNodes > ADDRMAN_GETADDR_MAX)
-        nNodes = ADDRMAN_GETADDR_MAX;
+    unsigned int nNodes = vRandom.size();
+    if (max_pct != 0) {
```
Comment (luke-jr) [inline src/addrman.cpp:485]: Seems more logical to use `>= 100` here?

File: src/addrman.cpp [hunk 9]
```diff
@@ -479,11 +479,15 @@ int CAddrMan::Check_()
...
+        nNodes = max_pct * nNodes / 100;
+    }
+    if (max_addresses != 0) {
```
Comment (luke-jr) [inline src/addrman.cpp:488]: Maybe just `std::numeric_limits<int>::max()`

File: src/addrman.h [hunk 10]
```diff
@@ -261,7 +255,7 @@ friend class CAddrManTest;
 #endif
 
     //! Select several addresses at once.
-    void GetAddr_(std::vector<CAddress> &vAddr) EXCLUSIVE_LOCKS_REQUIRED(cs);
+    void GetAddr_(std::vector<CAddress> &vAddr, unsigned int max_addresses, unsigned int max_pct) EXCLUSIVE_LOCKS_REQUIRED(cs);
```
Comment (luke-jr) [inline src/addrman.h:258]: `size_t` might be more appropriate for `max_addresses`

File: src/bench/addrman.cpp [hunk 11]
```diff
@@ -98,7 +98,7 @@ static void AddrManGetAddr(benchmark::Bench& bench)
     FillAddrMan(addrman);
 
     bench.run([&] {
-        const auto& addresses = addrman.GetAddr();
+        const auto& addresses = addrman.GetAddr(2500, 23);  // Compatibility with ADDRMAN_GETADDR_MAX_PCT and ADDRMAN_GETADDR_MAX
```
Comment (luke-jr) [inline src/bench/addrman.cpp:101]: This comment is confusing. I assume it's supposed to mean the original values thereof?

File: src/rpc/net.cpp [hunk 12]
```diff
@@ -727,7 +727,7 @@ static UniValue getnodeaddresses(const JSONRPCRequest& request)
             RPCHelpMan{"getnodeaddresses",
                 "\nReturn known addresses which can potentially be used to find new nodes in the network\n",
                 {
-                    {"count", RPCArg::Type::NUM, /* default */ "1", "How many addresses to return. Limited to the smaller of " + ToString(ADDRMAN_GETADDR_MAX) + " or " + ToString(ADDRMAN_GETADDR_MAX_PCT) + "% of all known addresses."},
+                    {"count", RPCArg::Type::NUM, /* default */ "1", "The max number of addresses to return. Specify 0 to return all known addresses."},
```
Comment (luke-jr) [inline src/rpc/net.cpp:730]: "maximum"

Comment (luke-jr) [inline src/rpc/net.cpp:730]: Would prefer -1 rather than 0 for no limit

File: src/rpc/net.cpp [hunk 13]
```diff
@@ -775,6 +773,54 @@ static UniValue getnodeaddresses(const JSONRPCRequest& request)
     return ret;
 }
 
+static UniValue addpeeraddress(const JSONRPCRequest& request)
```
Comment (luke-jr) [inline src/rpc/net.cpp:776]: IMO this should be moved to a separate PR, so not reviewing here.

**Review by luke-jr — CHANGES_REQUESTED** [review]

File: src/addrman.cpp (hunk 8 above)
Comment (jnewbery) [reply src/addrman.cpp:485]: I think both are fine. Calling with 0 to indicate no max seems fine to me too.

File: src/addrman.cpp (hunk 9 above)
Comment (jnewbery) [reply src/addrman.cpp:488]: as above

File: src/addrman.h (hunk 10 above)
Comment (jnewbery) [reply src/addrman.h:258]: ok, changed all of these to size_t

File: src/bench/addrman.cpp (hunk 11 above)
Comment (jnewbery) [reply src/bench/addrman.cpp:101]: Yes, so the values used in the bench stay the same. I've removed the comment since it's confusing.

File: src/rpc/net.cpp (hunk 12 above)
Comment (jnewbery) [reply src/rpc/net.cpp:730]: done

Comment (jnewbery) [reply src/rpc/net.cpp:730]: as above

File: src/rpc/net.cpp (hunk 13 above)
Comment (jnewbery) [reply src/rpc/net.cpp:776]: I'm keeping it in this PR since it already has reviews and concept ACKs.

**Comment by jnewbery** [comment]
Thanks for the review @luke-jr . I've addressed all of your comments.

**Comment by naumenkogs** [comment]
utACK 37a480e0cd94895b6051abef12d984ff74bdc4a3
I think both methods would be useful.

You updated rpc_net.py test, but there's also p2p_getaddr_cache.py and p2p_addr_relay.py which may benefit from new methods?

**Comment by jnewbery** [comment]
> there's also p2p_getaddr_cache.py and p2p_addr_relay.py which may benefit from new methods?

Of course. Those can be done in a follow-up.

**Comment by laanwj** [comment]
Code review and lightly manually tested ACK 37a480e0cd94895b6051abef12d984ff74bdc4a3
I found the default of 1 a bit confusing, but it's well documented so also no real problem with it.

**Comment by jnewbery** [comment]
Thanks @laanwj 

> I found the default of 1 a bit confusing (because the RPC name is plural), but it's well documented so also no real problem with it.

Yeah, I agree it's a bit confusing, but I didn't want to change the default in this PR.
